drivers: usb: Fix NXP driver for CDC test cases

mmahadevan108 opened the pull request:
The cdc_acm tests were failing. Remove a length check in the write function as this is handled by the HAL driver. Remove unused type member and improve the error checking.

Fixes #53636 , #52625

jfischer-no reviewed (changes requested):
Please split it in three commits, one for cleanup and error checking, second for the changes in usb_dc_ep_read_wait, and another for the changes in usb_dc_ep_write() to fix control IN transfers on IP3511.

mmahadevan108 committed:
drivers: usb: Cleanup MCUX driver

1. Remove unused ep_type structure member
2. Add more direction checks

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: usb: mcux: Remove unnecessary while loop

The while loop is changed to an if check.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: usb: mcux: Fix the write operation

1. Do not copy over the data to the local buffer,
it can be directly sent to the controller. The
cache is flushed before calling the HAL send
function. Also do not allocate a buffer from
the heap pool for the write operation.
2. Remove a length check as this is handled by
the HAL driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -490,37 +487,35 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
 	}
 
-	for (uint32_t n = 0; n < len_to_send; n++) {
+	for (uint32_t n = 0; n < data_len; n++) {
```
I have removed this copy.

jfischer-no replied:
Okay, but please check if there are requirements regarding alignment and memory type for IP3511 and EHCI controller, data can also point to an address in flash.

mmahadevan108 replied:
I have not seen any limitations. I have tested using various Zephyr USB samples such as `cdc_acm`, `mass`, `hid `etc.

mmahadevan108 commented:
> Please split it in three commits, one for cleanup and error checking, second for the changes in usb_dc_ep_read_wait, and another for the changes in usb_dc_ep_write() to fix control IN transfers on IP3511.

PR split into 3 commits. I have tested this PR for both the EHCI (RT1060) and IP3511 (LPC55S69) controllers

jfischer-no commented on drivers/usb/device/usb_dc_mcux.c:
```diff
@@ -486,8 +486,6 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
 		    const uint32_t data_len, uint32_t *const ret_bytes)
 {
 	uint8_t ep_abs_idx = EP_ABS_IDX(ep);
-	uint8_t *buffer = (uint8_t *)dev_state.eps[ep_abs_idx].block.data;
```
This intermediate buffer is not needed anymore. 
https://github.com/zephyrproject-rtos/zephyr/blob/e7443bb2035574ec1377b791032769e3d0a7f428/drivers/usb/device/usb_dc_mcux.c#L281
Should it still be allocated?

mmahadevan108 replied:
This is used for the receive.

jfischer-no replied:
It is allocated for both IN and OUT endpoints, but with this changes the driver does not need to allocate it for the IN endpoints.

mmahadevan108 commented:
@jfischer-no @desowin , can you take a look again.

jfischer-no reviewed (approved):
Maybe you want to reduce number of blocks as well,
`#define EP_BUF_NUMOF_BLOCKS	((NUM_OF_EP_MAX + 3) / 4)`

Otherwise LGTM, although I have not tested due to lack of time.
